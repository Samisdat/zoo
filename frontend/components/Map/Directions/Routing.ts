interface Route {
    length: number;
    nodes:number[];
    finished:boolean;
}

export interface Neighbours {
    [key:number]: number
}

export interface RoutingGraph{
    [key:number]: Neighbours
}

export class Routing{

    private shortest:number = Infinity

    private routes:Route[] = [];

    constructor(
        private graph:RoutingGraph,
        private start:number,
        private end:number
    ) {

        const route:Route = {
            length:0,
            nodes:[start],
            finished:false
        };

        this.routes.push(route);

        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();
        this.route();

        console.table(this.routes)

    }

    public getRoutes(){
        return this.routes;
    }

    public getShortest(){
        return this.shortest;
    }

    private route(){

        for(const route of this.routes){

            if(true === route.finished){
                continue;
            }

            const currentId = route.nodes[route.nodes.length - 1];

            const current = this.graph[currentId];

            let numberOfNeighbours = 0;

            for(const neighbour in current){

                const neighbourId = parseInt(neighbour, 10);

                // prevent back and forth
                if(true === route.nodes.includes(neighbourId)){

                    continue;
                }

                numberOfNeighbours += 1;

                if(1 === numberOfNeighbours){

                    route.nodes.push(neighbourId);
                    route.length += current[neighbour];

                }
                else{

                    const fork:Route = {
                        length:route.length,
                        finished:false,
                        nodes: [
                            ...route.nodes
                        ]
                    };

                    fork.nodes.push(neighbourId);
                    fork.length += current[neighbour];

                    this.routes.push(fork)

                }

            }

        }

        for(const route of this.routes){

            const last = route.nodes[route.nodes.length - 1];

            if(this.end === last){

                route.finished = true;

                if(route.length < this.shortest){
                    this.shortest = route.length;
                }

                //console.log('hit', route, this.shortest)

            }

            if(route.length > this.shortest){
                route.finished = true;
            }

        }

        const atLeastOneUnfinished = this.routes.filter((route)=>{
            return (false === route.finished);
        });

        console.log(atLeastOneUnfinished);

    }

}