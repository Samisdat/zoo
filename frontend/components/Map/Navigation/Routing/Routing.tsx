import React, {useEffect, useRef} from 'react';
import {Edge} from 'strapi-api/entity/edge/edge';
import * as d3 from 'd3';
import {useMap} from '../../Context/MapContext';


interface CurrentRouteProps {
    edges: Edge[];
}

export const Routing = ({edges}:CurrentRouteProps) => {

    const {
        state: {position, routing},
    } = useMap();

    const outerRef = useRef(null);
    const innerRef = useRef(null);

    useEffect(() => {

        if(!routing){
            return;
        }

        if('response' !== routing.type){
            return;
        }

        const route = routing.route;

        const current = edges.filter((edge)=>{

            if(edge.id === position.edgeId){
                return true;
            }

            return(
                route.nodes.includes(edge.startNode.id + '') &&
                route.nodes.includes(edge.endNode.id + '')
            )

        });

        const outerGroup = d3.select(outerRef.current);

        outerGroup.selectAll('path')
            .data(current)
            .join('path')
            .attr('d', (d)=>{
                return d.d;
            })
            .attr('fill', 'none')
            .attr('stroke', '#0A6001')
            .attr('stroke-width', '4px')
        ;

        const innerGroup = d3.select(innerRef.current);

        innerGroup.selectAll('path')
            .data(current)
            .join('path')
            .attr('d', (d)=>{
                return d.d;
            })
            .attr('fill', 'none')
            .attr('stroke', '#00a800')
            .attr('stroke-width', '2px')

    },[routing]);


    return (
        <React.Fragment>
            <g
                ref={outerRef}
            ></g>
            <g
                ref={innerRef}
            ></g>
        </React.Fragment>
    );

}
