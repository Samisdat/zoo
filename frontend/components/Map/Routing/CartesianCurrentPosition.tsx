import React, {useEffect, useRef} from 'react';
import * as d3 from 'd3';
import {MapTransformInterface, useMap} from "../Context/MapContext";
import {angle, svg} from "../../../constants";

export interface Coordinate{
    x:number;
    y:number;
}

export const rotateCords = (coordinate: Coordinate, degree:number):Coordinate => {

    if(180 !== degree){
        throw new Error('only 180 is implemented');
    }

    const center:Coordinate = {
        x: svg.width / 2,
        y: svg.height / 2
    }

    const rotated:Coordinate = {
        x: undefined,
        y: undefined
    };

    if(coordinate.x === center.x){
        rotated.x = center.x;
    }
    else if(coordinate.x > center.x){
        const x = coordinate.x - center.x;
        rotated.x = center.x - x;
    }
    else {
        const x = center.x - coordinate.x;
        rotated.x = center.x + x;
    }

    if(coordinate.y === center.y){
        rotated.y = center.y;
    }
    else if(coordinate.y > center.y){
        const y = coordinate.y - center.y;
        rotated.y = center.y - y;
    }
    else {
        const y = center.y - coordinate.y;
        rotated.y = center.y + y;
    }



    return rotated;

}

interface CartesianCurrentPositionProps{
    cartesianTransform:MapTransformInterface
}

export const CartesianCurrentPosition = ({cartesianTransform}:CartesianCurrentPositionProps) => {

    const {
        state: {position, projection, transform},
    } = useMap();

    const refPoint = useRef(null);
    const refPoint2 = useRef(null);

    useEffect(() => {
        

        if(!position || undefined === position.lat || undefined === position.lng){
            return;
        }
        if(!cartesianTransform){
            return;
        }

        const cartesianPos = projection([position.lng, position.lat]);

        const x = (cartesianPos[0] - cartesianTransform.x) / cartesianTransform.k;
        const y = (cartesianPos[1] - cartesianTransform.y) / cartesianTransform.k;

        const center = {
            y: (svg.height / 2),
            x: (svg.width / 2)
        };

        const rotate = `rotate(${ -1 * angle} ${center.x} ${center.y})`;

        const nodesGroup = d3.select(refPoint.current);
        let circle = nodesGroup.select('circle')

        if(!circle.node()){
            nodesGroup.append('circle');
            circle = nodesGroup.select('circle');
            circle.attr('id', 'CartesianCurrentPosition')
        }

        circle.attr('cx', function(d) {
            return x;
        })
        .attr('cy', function(d) {
            return y;
        })
        .attr('fill', (d, i)=>{

            return 'purple';

        })
        .attr('r', 20)

        .attr('transform', rotate)


        const nodesGroup2 = d3.select(refPoint2.current);
        let circle2 = nodesGroup2.select('circle')

        if(!circle2.node()){
            nodesGroup2.append('circle');
            circle2 = nodesGroup2.select('circle');
            circle2.attr('id', 'CartesianCurrentPosition_Without_Rotate')
        }

        const rotateCords1 = rotateCords({x,y},180);

        circle2.attr('cx', function(d) {
            return rotateCords1.x;
        })
        .attr('cy', function(d) {
            return rotateCords1.y;
        })
        .attr('fill', (d, i)=>{

            return 'yellow';

        })
        .attr('r', 30)
        .attr('opacity', .5)
        console.log(circle2.node())

    },[position, transform]);

    return (
        <React.Fragment>
            <g ref={refPoint}></g>
            <g ref={refPoint2}></g>
        </React.Fragment>
    );
}
