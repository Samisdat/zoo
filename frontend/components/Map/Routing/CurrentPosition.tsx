import React, {useEffect, useRef} from 'react';
import * as d3 from 'd3';
import {useMap} from "../Context/MapContext";

export const CurrentPosition = () => {

    const {
        state: {position},
    } = useMap();

    const ref = useRef(null);

    useEffect(() => {

        if(!position || undefined === position.x || undefined === position.y){
            return;
        }

        let radius = 30;

        if(position.fuzziness && 30 < position.fuzziness){
            radius = position.fuzziness;
        }

        d3.select(ref.current)
            .attr('cx', function(d) {
                return position.x;
            })
            .attr('cy', function(d) {
                return position.y;
            })
            .attr('r', radius)
        ;

    },[position]);

    return (
        <circle
            ref={ref}
            fill={'yellow'}
        />
    );
}
