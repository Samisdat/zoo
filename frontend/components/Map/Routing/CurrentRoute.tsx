import React, {useEffect, useRef} from 'react';
import {Edge} from "../../../strapi-api/entity/edge/edge";
import {makeStyles} from "@material-ui/core/styles";
import {edgeIdPrefix} from "../../../constants";
import * as d3 from "d3";
import {Route} from "./Dijkstra";
import {position} from "polished";
import {useMap} from "../Context/MapContext";

interface CurrentRouteProps {
    edges: Edge[];
}

const useStyles = makeStyles({
    outer:{
        fill: 'none',
        stroke: '#0A6001',
        strokeWidth: '4px',
    },
    inner:{
        fill: 'none',
        stroke: '#00a800',
        strokeWidth: '2px',
    }
});

export const CurrentRoute = ({edges}:CurrentRouteProps) => {

    const classes = useStyles();

    const {
        state: {position, routing},
    } = useMap();

    const outerRef = useRef(null);
    const innerRef = useRef(null);

    useEffect(() => {

        if(!routing){
            return;
        }

        if('response' !== routing.type){
            return;
        }

        const route = routing.route;

        const current = edges.filter((edge)=>{

            if(edge.id === position.edgeId){
                return true;
            }

            return(
                route.nodes.includes(edge.startNode.id + '') &&
                route.nodes.includes(edge.endNode.id + '')
            )

        });

        const outerGroup = d3.select(outerRef.current);

        outerGroup.selectAll('path')
            .data(current)
            .join('path')
            .attr('d', (d)=>{
                return d.d;
            })
            .attr('class', (d)=>{
                return `${classes.outer}`;
            })

        const innerGroup = d3.select(innerRef.current);

        innerGroup.selectAll('path')
            .data(current)
            .join('path')
            .attr('d', (d)=>{
                return d.d;
            })
            .attr('class', (d)=>{
                return `${classes.inner}`;
            })

    },[routing]);


    return (
        <React.Fragment>
            <g
                ref={outerRef}
            ></g>
            <g
                ref={innerRef}
            ></g>
        </React.Fragment>
    );

}
