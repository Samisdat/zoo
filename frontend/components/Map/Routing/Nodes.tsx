import React, {useEffect, useRef} from 'react';
import * as d3 from 'd3';

import {useMap} from "../Context/MapContext";
import {Node} from "../../../strapi-api/entity/node/node";

interface RoutingProperties {
    nodes: Node[];
}

export const Nodes = (props:RoutingProperties) => {

    const {
        state: {path, position},
    } = useMap();

    const ref = useRef(null);

    useEffect(() => {

        if(!path){
            return;
        }

        const nodesGroup = d3.select(ref.current);

        nodesGroup.selectAll('circle')
        .data(props.nodes)
        .join('circle')
        .attr('cx', function(d) {
                return Math.round(d.x);
        })
        .attr('cy', function(d) {
                return Math.round(d.y);
        })
        .attr('data-pos', function(d) {
                return d.x + '-' + d.y;
        })
        .attr('stroke', (d, i)=>{
                return '#000';
        })
        .attr('stroke-width', (d, i)=>{
                return 1;
        })
        .attr('vector-effect', (d, i)=>{
                return 'non-scaling-stroke'
                return 1;
        })
        .attr('fill', (d, i)=>{
                return 'red';
                /*
                if(true === route.nodes.includes(d.id + '')){
                        return 'red';
                }

                 */

                return 'lightgrey';

        })
        .attr('title', (d, i)=>{

                return d.id;

        })
        .attr('r', 5)
        .attr('d', path as any)
        ;

    },[path, position]);

    return (
        <g ref={ref} />
    );

}
