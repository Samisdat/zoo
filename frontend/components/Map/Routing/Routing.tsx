import React, {useEffect, useRef} from 'react';
import {Nodes} from "./Nodes";
import {Edges} from "./Edges";
import {Node} from "../../../strapi-api/entity/node/node";
import {Edge} from "../../../strapi-api/entity/edge/edge";
import {Dijkstra, Route, RoutingGraph} from "./Dijkstra";
import {MapTransformInterface} from "../Context/MapContext";

interface RoutingProperties {
    nodes: Node[];
    edges:Edge[];
    cartesianTransform:MapTransformInterface
}

const createGraph = (edges:Edge[]):RoutingGraph => {

    const graph:RoutingGraph = {}

    for(const edge of edges){


        if(undefined === graph[edge.startNode.id]){
            graph[edge.startNode.id] = [];
        }

        graph[edge.startNode.id].push({
            id:edge.endNode.id + '',
            distance:edge.edgeLength
        });

        if(undefined === graph[edge.endNode.id]){
            graph[edge.endNode.id] = [];
        }

        graph[edge.endNode.id].push({
            id:edge.startNode.id + '',
            distance:edge.edgeLength
        });

    }

    return graph;

};

export const Routing = (props:RoutingProperties) => {

    const refRouting = useRef(null);

    const graph = createGraph(props.edges);

    const [route, setRoute] = React.useState<Route>(undefined);

    useEffect(() => {

        const dijkstra = new Dijkstra(
            graph,
            399 + '',
            232 + '',
        );
        //return;
        setRoute(dijkstra.getShortestRoute());

    },[]);

    return (
        <g ref={refRouting}>
            <Edges
                edges={props.edges}
                route={route}
                cartesianTransform={props.cartesianTransform}
            />
            <Nodes
                nodes={props.nodes}
                route={route}
            />
        </g>
    );

}
