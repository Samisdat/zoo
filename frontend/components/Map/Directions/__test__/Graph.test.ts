import {Routing} from "../Routing";

const foo = {
    "0": {"1": 65.11895751953125, "2": 62.89460372924805},
    "1": {"0": 65.11895751953125, "2": 65.0198974609375, "51": 36.77623748779297},
    "2": {"0": 62.89460372924805, "1": 65.0198974609375, "60": 320.8688049316406},
    "3": {"4": 41.334041595458984, "5": 20.676132202148438, "15": 371.2860412597656},
    "4": {"3": 41.334041595458984, "179": 576.6243896484375},
    "5": {"3": 20.676132202148438, "8": 151.29229736328125},
    "6": {"7": 41.853275299072266},
    "7": {"6": 41.853275299072266},
    "8": {"5": 151.29229736328125, "9": 44.60441589355469, "10": 30.625276565551758},
    "9": {"8": 44.60441589355469, "10": 45.79067611694336, "59": 342.6955261230469},
    "10": {"8": 30.625276565551758, "9": 45.79067611694336, "21": 108.54898071289062},
    "11": {"12": 58.361053466796875, "13": 64.6456298828125, "169": 141.1676483154297},
    "12": {"11": 58.361053466796875, "13": 60.106624603271484, "166": 14.349714279174805},
    "13": {"11": 64.6456298828125, "12": 60.106624603271484, "173": 254.48500061035156},
    "14": {"15": 119.18959045410156, "17": 59.784584045410156, "88": 89.30005645751953},
    "15": {"3": 371.2860412597656, "14": 119.18959045410156, "16": 51.497528076171875},
    "16": {"15": 51.497528076171875, "17": 109.72776794433594},
    "17": {"14": 59.784584045410156, "16": 109.72776794433594, "158": 98.39749145507812},
    "18": {"19": 30.522058486938477, "21": 24.941450119018555, "58": 390.3686828613281},
    "19": {"18": 30.522058486938477, "20": 26.02042579650879, "218": 230.223388671875},
    "20": {"19": 26.02042579650879},
    "21": {"10": 108.54898071289062, "18": 24.941450119018555},
    "22": {"23": 35.316139221191406, "27": 7.318266868591309},
    "23": {"22": 35.316139221191406, "24": 28.00531768798828, "215": 315.5186767578125},
    "24": {"23": 28.00531768798828, "25": 34.85498809814453, "45": 104.8480453491211},
    "25": {"24": 34.85498809814453},
    "26": {"27": 80.24699401855469, "28": 81.03394317626953, "29": 73.70817565917969, "164": 231.34686279296875},
    "27": {"22": 7.318266868591309, "26": 80.24699401855469, "28": 75.00874328613281, "29": 58.277626037597656},
    "28": {"26": 81.03394317626953, "27": 75.00874328613281, "29": 95.59053039550781, "79": 26.71684455871582},
    "29": {"26": 73.70817565917969, "27": 58.277626037597656, "28": 95.59053039550781, "43": 55.39665603637695},
    "30": {"31": 19.348934173583984, "32": 17.194290161132812},
    "31": {"30": 19.348934173583984, "32": 13.12743091583252, "150": 167.736572265625},
    "32": {"30": 17.194290161132812, "31": 13.12743091583252},
    "33": {"34": 41.873390197753906, "36": 31.090126037597656},
    "34": {"33": 41.873390197753906, "35": 33.60470199584961, "39": 90.0361328125},
    "35": {"34": 33.60470199584961},
    "36": {"33": 31.090126037597656, "54": 185.5693359375},
    "37": {"38": 65.78197479248047, "39": 61.053001403808594, "44": 70.40023803710938},
    "38": {"37": 65.78197479248047, "39": 67.24959564208984, "188": 129.23394775390625},
    "39": {"34": 90.0361328125, "37": 61.053001403808594, "38": 67.24959564208984},
    "40": {"41": 48.303707122802734, "42": 54.196632385253906, "207": 138.30226135253906},
    "41": {"40": 48.303707122802734, "42": 44.4422607421875, "66": 129.1467742919922},
    "42": {"40": 54.196632385253906, "41": 44.4422607421875, "67": 26.558069229125977},
    "43": {"29": 55.39665603637695, "44": 33.8682861328125, "45": 28.36673355102539},
    "44": {"37": 70.40023803710938, "43": 33.8682861328125, "45": 28.269760131835938},
    "45": {"24": 104.8480453491211, "43": 28.36673355102539, "44": 28.269760131835938},
    "46": {"47": 47.55728530883789, "48": 55.312965393066406, "49": 69.3797607421875, "86": 178.6003875732422},
    "47": {"46": 47.55728530883789, "48": 57.72491455078125, "50": 71.2511215209961, "89": 111.93973541259766},
    "48": {"46": 55.312965393066406, "47": 57.72491455078125, "49": 37.99482727050781, "216": 82.38323974609375},
    "49": {"46": 69.3797607421875, "48": 37.99482727050781, "104": 259.37158203125},
    "50": {"47": 71.2511215209961},
    "51": {"1": 36.77623748779297, "52": 35.38310241699219, "53": 37.28236389160156},
    "52": {"51": 35.38310241699219, "53": 32.357818603515625},
    "53": {"51": 37.28236389160156, "52": 32.357818603515625},
    "54": {"36": 185.5693359375, "55": 48.511741638183594, "56": 26.161623001098633},
    "55": {"54": 48.511741638183594, "56": 35.50856018066406, "203": 180.49090576171875},
    "56": {"54": 26.161623001098633, "55": 35.50856018066406, "70": 190.3048095703125},
    "57": {"58": 73.13946533203125, "59": 52.138999938964844, "157": 174.1454315185547},
    "58": {"18": 390.3686828613281, "57": 73.13946533203125, "59": 47.986568450927734},
    "59": {"9": 342.6955261230469, "57": 52.138999938964844, "58": 47.986568450927734},
    "60": {"2": 320.8688049316406, "61": 80.5591812133789, "62": 73.8965072631836},
    "61": {"60": 80.5591812133789, "217": 376.755615234375},
    "62": {"60": 73.8965072631836, "63": 56.29262924194336},
    "63": {"62": 56.29262924194336},
    "64": {"65": 22.60488510131836, "66": 16.75212860107422},
    "65": {"64": 22.60488510131836, "66": 25.23563003540039, "149": 118.78520965576172},
    "66": {"41": 129.1467742919922, "64": 16.75212860107422, "65": 25.23563003540039},
    "67": {"42": 26.558069229125977, "68": 47.60432815551758, "70": 25.672788619995117, "72": 38.43745803833008},
    "68": {"67": 47.60432815551758, "70": 36.2885856628418},
    "69": {"70": 21.47747039794922, "71": 26.771703720092773},
    "70": {"56": 190.3048095703125, "67": 25.672788619995117, "68": 36.2885856628418, "69": 21.47747039794922},
    "71": {"69": 26.771703720092773, "83": 109.79655456542969},
    "72": {"67": 38.43745803833008, "208": 36.8303337097168},
    "73": {"74": 53.00328063964844, "75": 58.43726348876953, "76": 43.70008850097656, "209": 46.934696197509766},
    "74": {"73": 53.00328063964844, "76": 48.928932189941406, "182": 481.69415283203125},
    "75": {"73": 58.43726348876953, "76": 60.75518798828125, "77": 41.068267822265625, "147": 691.1405029296875},
    "76": {"73": 43.70008850097656, "74": 48.928932189941406, "75": 60.75518798828125, "210": 106.47720336914062},
    "77": {"75": 41.068267822265625},
    "78": {"79": 36.81244659423828, "80": 31.98369026184082, "194": 174.73574829101562},
    "79": {"28": 26.71684455871582, "78": 36.81244659423828, "80": 40.048797607421875},
    "80": {"78": 31.98369026184082, "79": 40.048797607421875},
    "81": {"82": 46.188175201416016, "83": 50.272396087646484, "148": 50.16436767578125},
    "82": {"81": 46.188175201416016, "83": 33.8101806640625, "204": 101.38343048095703},
    "83": {"71": 109.79655456542969, "81": 50.272396087646484, "82": 33.8101806640625},
    "84": {"85": 81.5380859375, "86": 59.54883575439453, "198": 193.52301025390625},
    "85": {"84": 81.5380859375, "86": 90.49485778808594},
    "86": {"46": 178.6003875732422, "84": 59.54883575439453, "85": 90.49485778808594},
    "87": {"88": 61.74336624145508, "89": 36.5721549987793},
    "88": {"14": 89.30005645751953, "87": 61.74336624145508, "89": 56.04325866699219, "90": 47.85703659057617},
    "89": {"47": 111.93973541259766, "87": 36.5721549987793, "88": 56.04325866699219, "90": 74.76243591308594},
    "90": {"88": 47.85703659057617, "89": 74.76243591308594, "91": 75.04541015625, "156": 236.726318359375},
    "91": {"90": 75.04541015625, "96": 384.57562255859375},
    "92": {"93": 44.28453826904297, "99": 87.517578125},
    "93": {"92": 44.28453826904297},
    "94": {"95": 32.74489212036133},
    "95": {"94": 32.74489212036133},
    "96": {"91": 384.57562255859375, "97": 39.85813903808594},
    "97": {"96": 39.85813903808594, "102": 620.6104736328125},
    "98": {"99": 51.608642578125, "100": 42.85279083251953, "101": 40.07380676269531, "110": 172.0262908935547},
    "99": {"92": 87.517578125, "98": 51.608642578125, "101": 31.030548095703125, "102": 53.30640411376953},
    "100": {"98": 42.85279083251953},
    "101": {"98": 40.07380676269531, "99": 31.030548095703125, "103": 47.14106369018555, "214": 78.6229248046875},
    "102": {"97": 620.6104736328125, "99": 53.30640411376953},
    "103": {"101": 47.14106369018555},
    "104": {"49": 259.37158203125, "105": 24.236555099487305, "106": 17.198087692260742},
    "105": {"104": 24.236555099487305, "106": 22.32256507873535},
    "106": {"104": 17.198087692260742, "105": 22.32256507873535, "216": 277.47381591796875},
    "107": {"108": 25.589984893798828, "109": 21.988142013549805, "110": 26.31149673461914, "114": 87.90985870361328},
    "108": {"107": 25.589984893798828, "110": 19.441238403320312},
    "109": {"107": 21.988142013549805},
    "110": {"98": 172.0262908935547, "107": 26.31149673461914, "108": 19.441238403320312},
    "111": {"112": 22.884065628051758, "213": 94.60161590576172},
    "112": {"111": 22.884065628051758, "222": 11.260282516479492},
    "113": {"114": 27.92621421813965, "115": 32.524864196777344, "118": 41.492042541503906},
    "114": {"107": 87.90985870361328, "113": 27.92621421813965, "115": 23.373565673828125},
    "115": {"113": 32.524864196777344, "114": 23.373565673828125, "116": 29.85503578186035, "137": 95.70372772216797},
    "116": {"115": 29.85503578186035, "119": 77.14482116699219},
    "117": {"118": 16.20017433166504, "119": 14.731640815734863, "121": 83.61320495605469},
    "118": {"113": 41.492042541503906, "117": 16.20017433166504, "119": 16.392316818237305},
    "119": {"116": 77.14482116699219, "117": 14.731640815734863, "118": 16.392316818237305},
    "120": {"121": 12.876449584960938, "122": 15.033510208129883, "136": 117.19429779052734},
    "121": {"117": 83.61320495605469, "120": 12.876449584960938, "122": 13.623379707336426},
    "122": {"120": 15.033510208129883, "121": 13.623379707336426, "221": 10.741774559020996},
    "123": {"124": 46.76296615600586, "125": 29.9621524810791, "139": 482.2847595214844},
    "124": {"123": 46.76296615600586, "125": 35.53739929199219, "128": 36.30426025390625},
    "125": {"123": 29.9621524810791, "124": 35.53739929199219, "125": 43.02780532836914},
    "126": {"127": 46.49882888793945, "132": 8.344319343566895},
    "127": {"126": 46.49882888793945, "128": 97.20683288574219, "135": 10.36408805847168},
    "128": {"124": 36.30426025390625, "127": 97.20683288574219, "129": 61.277976989746094},
    "129": {"128": 61.277976989746094, "130": 22.14677619934082, "132": 34.41434097290039},
    "130": {"129": 22.14677619934082, "131": 29.777902603149414},
    "131": {"130": 29.777902603149414, "212": 5.825003147125244},
    "132": {"126": 8.344319343566895, "129": 34.41434097290039},
    "133": {"134": 56.20145797729492},
    "134": {"133": 56.20145797729492},
    "135": {"127": 10.36408805847168, "136": 42.57086944580078, "137": 32.481475830078125},
    "136": {"120": 117.19429779052734, "135": 42.57086944580078, "137": 45.27263641357422},
    "137": {"115": 95.70372772216797, "135": 32.481475830078125, "136": 45.27263641357422},
    "138": {"139": 21.76820945739746, "140": 18.348947525024414, "142": 111.8048095703125},
    "139": {"123": 482.2847595214844, "138": 21.76820945739746, "140": 16.807907104492188},
    "140": {"138": 18.348947525024414, "139": 16.807907104492188, "220": 5.766439914703369},
    "141": {"142": 19.738309860229492, "143": 16.21772003173828, "145": 7.967153072357178},
    "142": {"138": 111.8048095703125, "141": 19.738309860229492, "143": 16.202383041381836},
    "143": {"141": 16.21772003173828, "142": 16.202383041381836, "219": 19.07706069946289},
    "144": {"145": 19.76683235168457, "146": 15.382694244384766, "211": 68.29652404785156},
    "145": {"141": 7.967153072357178, "144": 19.76683235168457, "146": 14.487287521362305},
    "146": {"144": 15.382694244384766, "145": 14.487287521362305, "153": 13.478615760803223},
    "147": {"75": 691.1405029296875, "148": 60.4193000793457, "149": 91.34190368652344, "150": 99.23204040527344},
    "148": {"81": 50.16436767578125, "147": 60.4193000793457, "149": 93.38549041748047, "150": 100.21659851074219},
    "149": {"65": 118.78520965576172, "147": 91.34190368652344, "148": 93.38549041748047, "150": 65.09156799316406},
    "150": {"31": 167.736572265625, "147": 99.23204040527344, "148": 100.21659851074219, "149": 65.09156799316406},
    "151": {
        "152": 51.862060546875,
        "153": 83.37330627441406,
        "154": 128.69525146484375,
        "155": 116.30839538574219,
        "167": 181.26065063476562
    },
    "152": {"151": 51.862060546875, "190": 78.67888641357422},
    "153": {"146": 13.478615760803223, "151": 83.37330627441406},
    "154": {"151": 128.69525146484375, "155": 122.82051849365234},
    "155": {"151": 116.30839538574219, "154": 122.82051849365234, "191": 207.69337463378906},
    "156": {"90": 236.726318359375, "157": 25.424842834472656},
    "157": {"57": 174.1454315185547, "156": 25.424842834472656, "158": 15.525625228881836},
    "158": {"17": 98.39749145507812, "157": 15.525625228881836, "159": 25.193161010742188},
    "159": {"158": 25.193161010742188},
    "160": {"161": 47.0754280090332, "162": 39.520103454589844},
    "161": {"160": 47.0754280090332, "162": 42.69454574584961, "165": 78.33828735351562},
    "162": {"160": 39.520103454589844, "161": 42.69454574584961, "195": 282.6112060546875},
    "163": {"164": 48.58611297607422, "165": 47.674468994140625, "168": 117.40644836425781},
    "164": {"26": 231.34686279296875, "163": 48.58611297607422, "165": 45.62703323364258},
    "165": {"161": 78.33828735351562, "163": 47.674468994140625, "164": 45.62703323364258},
    "166": {"12": 14.349714279174805, "167": 75.8837661743164, "168": 66.4890365600586},
    "167": {"151": 181.26065063476562, "166": 75.8837661743164, "168": 54.34542465209961},
    "168": {"163": 117.40644836425781, "166": 66.4890365600586, "167": 54.34542465209961},
    "169": {"11": 141.1676483154297, "170": 93.41513061523438, "171": 63.547786712646484},
    "170": {"169": 93.41513061523438, "171": 68.5495376586914, "187": 45.23902893066406},
    "171": {"169": 63.547786712646484, "170": 68.5495376586914, "189": 120.1754150390625},
    "172": {"173": 29.43326187133789, "174": 26.45542335510254, "176": 151.3635711669922},
    "173": {"13": 254.48500061035156, "172": 29.43326187133789, "174": 27.076425552368164},
    "174": {"172": 26.45542335510254, "173": 27.076425552368164},
    "175": {"176": 48.00552749633789, "177": 45.61604690551758, "178": 26.44529914855957},
    "176": {"172": 151.3635711669922, "175": 48.00552749633789, "177": 65.35944366455078},
    "177": {"175": 45.61604690551758, "176": 65.35944366455078, "186": 230.02955627441406},
    "178": {"175": 26.44529914855957},
    "179": {"4": 576.6243896484375},
    "180": {"181": 305.6116638183594, "182": 160.4023895263672},
    "181": {"180": 305.6116638183594},
    "182": {"74": 481.69415283203125, "180": 160.4023895263672},
    "183": {"184": 84.04998779296875},
    "184": {"183": 84.04998779296875},
    "185": {"186": 147.69143676757812},
    "186": {"177": 230.02955627441406, "185": 147.69143676757812},
    "187": {"170": 45.23902893066406},
    "188": {"38": 129.23394775390625},
    "189": {"171": 120.1754150390625},
    "190": {"152": 78.67888641357422},
    "191": {"155": 207.69337463378906},
    "192": {"193": 15.170112609863281},
    "193": {"192": 15.170112609863281},
    "194": {"78": 174.73574829101562},
    "195": {"162": 282.6112060546875},
    "196": {"197": 84.14212799072266},
    "197": {"196": 84.14212799072266},
    "198": {"84": 193.52301025390625},
    "199": {"200": 158.23629760742188},
    "200": {"199": 158.23629760742188},
    "201": {"202": 107.25173950195312},
    "202": {"201": 107.25173950195312},
    "203": {"55": 180.49090576171875},
    "204": {"82": 101.38343048095703},
    "205": {"206": 249.84860229492188},
    "206": {"205": 249.84860229492188},
    "207": {"40": 138.30226135253906},
    "208": {"72": 36.8303337097168},
    "209": {"73": 46.934696197509766},
    "210": {"76": 106.47720336914062},
    "211": {"144": 68.29652404785156},
    "212": {"131": 5.825003147125244},
    "213": {"111": 94.60161590576172},
    "214": {"101": 78.6229248046875},
    "215": {"23": 315.5186767578125},
    "216": {"48": 82.38323974609375, "106": 277.47381591796875},
    "217": {"61": 376.755615234375},
    "218": {"19": 230.223388671875},
    "219": {"143": 19.07706069946289},
    "220": {"140": 5.766439914703369},
    "221": {"122": 10.741774559020996},
    "222": {"112": 11.260282516479492},
    "223": {"224": 6.900559425354004},
    "224": {"223": 6.900559425354004},
    "225": {"226": 7.688976764678955},
    "226": {"225": 7.688976764678955}
};

describe('Graph', () => {

    test('data structure', async () => {

        const routing = new Routing(
            foo,
            171,
            11
        );

        console.log(routing.getShortest());

        expect(routing.getRoutes()).toMatchSnapshot()

    });

});



