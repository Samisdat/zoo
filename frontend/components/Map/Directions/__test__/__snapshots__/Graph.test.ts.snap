// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph data structure 1`] = `
Array [
  Object {
    "finished": true,
    "length": 204.71543502807617,
    "nodes": Array [
      171,
      169,
      11,
    ],
  },
  Object {
    "finished": false,
    "length": 177.33635330200195,
    "nodes": Array [
      171,
      169,
      170,
      187,
    ],
  },
  Object {
    "finished": false,
    "length": 183.72320175170898,
    "nodes": Array [
      171,
      169,
      189,
    ],
  },
  Object {
    "finished": true,
    "length": 343.3695945739746,
    "nodes": Array [
      171,
      169,
      11,
      170,
      187,
    ],
  },
]
`;
